PPU async Inductor workers reinitialize CUDA after fork

## Summary

FlagTree's PPU backend fails in PyTorch Inductor's asynchronous compile workers after the parent process has initialized CUDA. The worker calls `PPUDriver.get_active_torch_device()`, which calls `torch.cuda.current_device()` and attempts to initialize CUDA in a forked subprocess.

Serial compilation with `TORCHINDUCTOR_COMPILE_THREADS=1` avoids the worker process and makes the full Torch-FL compile suite pass.

## Environment

- Hardware: T-Head PPU, 16 visible devices
- FlagTree commit: `52678a8b781d2a3e36fc1a77fe74a9aa23c5ca4b`
- FlagTree package: `0.6.0+ppu.git52678a8b`
- Triton API version: `3.6.0`
- Build: `FLAGTREE_BACKEND=ppu MAX_JOBS=16 pip install . --no-build-isolation`
- PyTorch: `2.10.0`, CUDA compatibility version `13.0`
- Consumer: Torch-FL `2e00b393cf80088706b460a187aef185d3a283f4`
- Python: `3.12.13`

FlagTree detection is correct:

```text
triton 3.6.0
triton._flagtree_spec present
FLAGTREE_BACKEND ppu
```

## Reproduction

Run Torch-FL's compile suite with the default Inductor worker count (`32` in this environment) and a fresh cache:

```bash
PYTHONPATH=/tmp/flagtree-ppu-target \
FLAGOS_DISABLE_CUDA_ASSETS=1 \
FLAGOS_USE_FLAGTREE=1 \
TORCHINDUCTOR_CACHE_DIR=/tmp/torchinductor-flagtree-ppu-default \
pytest tests/integration/test_compile.py -v --tb=short
```

Observed result:

```text
13 passed, 2 failed

FAILED tests/integration/test_compile.py::test_compile_backward
FAILED tests/integration/test_compile.py::test_compile_dtypes[dtype1]
```

Both failures end with:

```text
File "triton/compiler/hint_manager.py", line 127, in hint_get_flagtree_backend
  device = driver.active.get_active_torch_device()
File "triton/backends/ppu/driver.py", line 682, in get_active_torch_device
  return torch.device("cuda", self.get_current_device())
File "torch/cuda/__init__.py", line 1094, in current_device
  _lazy_init()
RuntimeError: Cannot re-initialize CUDA in forked subprocess. To use CUDA with
multiprocessing, you must use the 'spawn' start method
```

The failure is reproducible with a clean Inductor cache. Which tests fail depends on when a new worker-side compilation is needed, but backward and FP16 consistently exercise it in the full suite.

## Workaround and control results

Force in-process/serial compilation:

```bash
PYTHONPATH=/tmp/flagtree-ppu-target \
FLAGOS_DISABLE_CUDA_ASSETS=1 \
FLAGOS_USE_FLAGTREE=1 \
TORCHINDUCTOR_COMPILE_THREADS=1 \
pytest tests/integration/test_compile.py -v --tb=short
```

Result:

```text
15 passed, 14 warnings in 5.22s
```

The dedicated FlagTree end-to-end test also passes:

```text
tests/integration/test_compile.py::test_flagtree_compiles_correct_results PASSED
```

A compiled PPU smoke test matches eager and returns on `flagos:0`:

```text
triton 3.6.0
flagtree_active True
flagtree_backend ppu
device flagos:0
match True
```

This indicates that PPU lowering, code generation, launch, FP16, and backward are functional; the failure is specific to device discovery in the asynchronous compile worker.

## Root cause

`triton/compiler/hint_manager.py::hint_get_flagtree_backend()` asks the active driver for a torch device. The PPU implementation is:

```python
def get_active_torch_device(self):
    import torch
    return torch.device("cuda", self.get_current_device())
```

`self.get_current_device()` is `torch.cuda.current_device`. In a worker created after CUDA initialization in the parent, this triggers PyTorch's fork-safety guard.

The PPU driver should avoid CUDA runtime initialization when the hint manager only needs a device/backend hint, or the compile worker must use a CUDA-safe process start/initialization path.

## Expected behavior

FlagTree PPU compilation should work with Inductor's default asynchronous compilation configuration after the parent has initialized a PPU/CUDA context. Calling `get_active_torch_device()` in the worker should not attempt to reinitialize CUDA after fork.

## Possible directions

- Make the PPU hint path derive the backend/device without calling `torch.cuda.current_device()` in the worker.
- Pass the selected device index from the parent into compile workers.
- Ensure the relevant worker uses a CUDA-safe start method.
- Add a regression test that initializes CUDA in the parent and compiles a new PPU kernel asynchronously.

--- a/python/triton/compiler/hint_manager.py
+++ b/python/triton/compiler/hint_manager.py
@@ -23,6 +23,8 @@
 import sys
 import importlib
 
+from triton._flagtree_backend import FLAGTREE_BACKEND
+
 
 class BaseHintHandler:
     # dynamicly find method
@@ -62,7 +64,7 @@ class HintManager:
         self.handler = self._load_handler(backend_name)
 
     def _load_handler(self, backend):
-        if backend == 'npu':
+        if backend == 'ascend':
             try:
                 module = importlib.import_module("triton.backends.ascend.ascend_hint_handler")
                 return module.AscendHintHandler()
@@ -76,7 +78,7 @@ class HintManager:
             except ImportError as e:
                 print(f"[FlagTree] Warning: Failed to load aipu Hint Handler: {e}", file=sys.stderr)
                 return BaseHintHandler()
-        elif backend == 'cuda':
+        elif backend == 'nvidia':
             try:
                 module = importlib.import_module("triton.backends.nvidia.nvidia_hint_handler")
                 return module.NvidiaHintHandler()
@@ -94,66 +96,6 @@ class HintManager:
             return BaseHintHandler()
 
 
-# supported backend with matched version
-SUPPORTED_BACKENDS = ["aipu", "npu", "cuda", "sunrise"]
-
-# TODO : npu will have conflicts if more backend involved
-# mapping name
-BACKEND_ALIASES = {
-    "ascend": "npu",
-    "huawei": "npu",
-    "nvidia": "cuda",
-    # sunrise: GPUTarget backend name is "tang", torch device type is "ptpu".
-    "tang": "sunrise",
-    "ptpu": "sunrise",
-}
-
-
-def normalize_backend_name(name: str) -> str:
-    if not name:
-        return ""
-    name = name.lower()
-    return BACKEND_ALIASES.get(name, name)
-
-
-def hint_get_flagtree_backend() -> str:
-    detected_backend = ""
-
-    # Priority 1: Triton Driver
-    try:
-        import torch
-        from triton.runtime import driver
-        if hasattr(driver, 'active') and hasattr(driver.active, 'get_active_torch_device'):
-            device = driver.active.get_active_torch_device()
-            if isinstance(device, torch.device):
-                detected_backend = device.type
-            # unimplemented support
-            elif isinstance(device, str):
-                detected_backend = device
-    except ImportError:
-        return ""
-
-    # TODO : some backend may not support priority 1, so keep priority 2 is necessary
-    # Priority 2: Torch Global State
-    if not detected_backend:
-        check_priority = ["aipu", "npu", "cuda"]
-
-        # 3. parse according to benefit
-        for candidate in check_priority:
-            module = getattr(torch, candidate, None)
-            if module and hasattr(module, "is_available") and module.is_available():
-                detected_backend = candidate
-                break
-
-    # (Normalization and Validation)
-    canonical_backend = normalize_backend_name(detected_backend)
-
-    if not canonical_backend or canonical_backend not in SUPPORTED_BACKENDS:
-        return ""
-
-    return canonical_backend
-
-
 # lazy load after first call hint trigger
 _global_hint_manager = None
 
@@ -162,5 +104,7 @@ def hint_trigger(hook_name, *args, **kwargs):
     global _global_hint_manager
 
     if _global_hint_manager is None:
-        _global_hint_manager = HintManager(hint_get_flagtree_backend())
+        # NVIDIA builds have no FlagTree backend marker.
+        backend_name = FLAGTREE_BACKEND or "nvidia"
+        _global_hint_manager = HintManager(backend_name)
     return _global_hint_manager.handler.trigger(hook_name, *args, **kwargs)

